eisbuk-admin/functions/src/utils.ts

```diff
@@ -186,3 +186,75 @@ export const sendRequest = <
     }
     req.end();
   });
+
+type ErrorCode = functions.https.FunctionsErrorCode;
+
+/** */
+export class EisbukHttpsError extends functions.https.HttpsError {
+  /**
+   * A wrapper around `functions.https.HttpsError` used to log errors
+   * to the functions console before returning gRPC error to the client
+   * @param code `functions.https.FunctionsErrorCode`
+   * @param message error message
+   * @param details error details
+   */
+  constructor(code: ErrorCode, message: string, details?: unknown) {
+    functions.logger.error(code, message, details);
+    super(code, message, details);
+  }
+}
+
+/**
+ * A function wrapper used to enforce a timeout on async
+ * operations with possibly long exectution period
+ * @param fn an async function we want to execute within timeout boundary
+ * @returns `fn`'s resolved value
+ */
+export const runWithTimeout = async <T extends any>(
+  fn: () => Promise<T>,
+  { timeout } = { timeout: 5000 }
+): Promise<T> => {
+  // set a timeout boundry for a function
+  const timeoutBoundary = setTimeout(() => {
+    throw new functions.https.HttpsError("aborted", HTTPSErrors.TimedOut);
+  }, timeout);
+
+  const res = await fn();
+
+  // clear timeout function (the `fn` has resolved within the timeout boundry)
+  clearTimeout(timeoutBoundary);
+
+  return res;
+};
+
+/**
+ * Checks payload (or any other record structure) for required fields.
+ * Throws error if one or more of the fields aren't present
+ * @param payload record to validate
+ * @param requiredFields array of field names (keys)
+ * @param errorMessage (optional) customer error message on failure, falls back to
+ * `HTTPSErrors.MissingParameter`
+ */
+export const checkRequiredFields = (
+  payload: Record<string, any> | undefined | null,
+  requiredFields: string[],
+  errorMessage: string = HTTPSErrors.MissingParameter
+): void | never => {
+  if (!payload) {
+    throw new EisbukHttpsError("invalid-argument", HTTPSErrors.NoPayload);
+  }
+
+  const missingFields: string[] = [];
+
+  Object.keys(requiredFields).forEach((field) => {
+    if (!payload[field]) {
+      missingFields.push(field);
+    }
+  });
+
+  if (missingFields.length) {
+    throw new EisbukHttpsError("invalid-argument", errorMessage, {
```

Comment:
I believe this should be `new EisbukHttpsError("invalid-argument", errorMessage, missingFields)`.

Reply:
I dispute this...kinda. `details` is fixed to `functions.https.HttpsError` creator (which is the actual type of `EisbukHttpsError`), so doing `{ missingFields: .... }` provides more detail in why the error was thrown (otherwise we would get `details: ["field-1", "field-2"]`) which, in my opinion is not as descriptive.

The problem stands that I should fix the way the missing fields are reported: should be field names, not indexes (ofc) which I think has to do with passing of missing fields into `missingFields` array, rather than error construction.

Reply:
Agreed! I misread the code and commented thinking it was `{... missingFields}` but it's actually `{missingFields}`.